### DISCUSSION TITLE
X509 name with specified encoding

### DISCUSSION THREAD

**Discussion by microshine** [PeculiarVentures/x509#27] [Ideas]
For some cases, we need to determine what encoding we should use for name creation. I'd like to discuss here and decide how we could extend the `Name` class to support this feature

The current implementation uses harden coded logic and doesn't allow doing it dynamically. `IA5String` encoding is used for `E` and `DC` names, and `PrintableString` or `UTF8String` encoding depends on incoming string.

```ts
if (type === this.getName("E") || type === this.getName("DC")) {
  attr.value.ia5String = value;
} else {
  // Use Utf8String for non ASCII strings
  if (Name.isASCII(value)) {
    attr.value.printableString = value;
  } else {
    attr.value.utf8String = value;
  }
}
```

**Comment by microshine** [comment]
I've got 3 ideas how we could do it

1. To use type prefixes for string representation of distinguished names

it doesn't match [RFC2253](https://datatracker.ietf.org/doc/html/rfc2253), but allows specifying any encoding for any attribute

```ts
const name = new Name("CN=p:Some name, O=u8:Home");
```

where prefixes are:
```
t  TeletexString
p  PrintableString
u  UniversalString
u8 UTF8String
b  BMPString
```

3. To use encoding profile for `Name` creation

It doesn't allow specifying multiple encoding for one attribute type (eg `CN=p:Test, CN=u8:Value`)

```ts
const name = new Name("CN=Some name, O=Home", {
  CN: NameEncoding.PrintableString,
  O: NameEncoding.UTF8String,
});
```

5. To extend and improve `JsonName`

```ts
const name = new x509.Name([
  {
    CN: [
      {
        printableString: "Some name",
      },
    ]
  },
  {
    O: [
      {
        utfString: "Home",
      },
    ]
  },
]);
```

**Reply by turon** [reply]
Per discussion in #25, my vote would be option 5 (the last of the three proposals).
The JSON API is generally provided for lower level control, so seems like an appropriate place to add options for specifying specific string encoding.

**Reply by turon** [reply]
Also, on option 1, wouldn't a more "typescript" style be: `const name = new Name("CN:p=Some name, O:u8=Home");`?

**Reply by microshine** [reply]
> const name = new Name("CN:p=Some name, O:u8=Home");

Looks good. This format doesn't require escaping ':' char in names
